pkg/gather/azure/azure.go

```diff
@@ -0,0 +1,293 @@
+				continue
+			}
+			if keyResults.Keys != nil {
+				for _, key := range *keyResults.Keys {
+					if key.Value != nil {
+						credential, err := azblob.NewSharedKeyCredential(accountName, to.String(key.Value))
+						if err != nil {
+							g.logger.Debugf("failed to get shared key: %s", err.Error())
+							continue
+						}
+						credentials = append(credentials, credential)
+					}
+				}
+			}
+		}
+	}
+
+	return credentials, nil
+}
+
+func getVirtualMachines(ctx context.Context, g *Gather) ([]compute.VirtualMachine, error) {
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	vmsPage, err := g.virtualMachinesClient.List(ctx, g.resourceGroupName)
+	if err != nil {
+		return nil, err
+	}
+
+	var virtualMachines []compute.VirtualMachine
+	for ; vmsPage.NotDone(); err = vmsPage.NextWithContext(ctx) {
+		if err != nil {
+			g.logger.Debugf("failed to get vm: %s", err.Error())
+			continue
+		}
+		for _, virtualMachine := range vmsPage.Values() {
+			virtualMachines = append(virtualMachines, virtualMachine)
+		}
+	}
+
+	return virtualMachines, nil
+}
+
+func getBootDiagnostics(ctx context.Context, virtualMachines []compute.VirtualMachine, g *Gather) []*compute.BootDiagnosticsInstanceView {
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	var bootDiagnostics []*compute.BootDiagnosticsInstanceView
+	for _, vm := range virtualMachines {
+		if vm.VirtualMachineProperties.DiagnosticsProfile != nil &&
+			vm.VirtualMachineProperties.DiagnosticsProfile.BootDiagnostics != nil &&
+			to.Bool(vm.VirtualMachineProperties.DiagnosticsProfile.BootDiagnostics.Enabled) {
+			instanceView, err := g.virtualMachinesClient.InstanceView(ctx, g.resourceGroupName, to.String(vm.Name))
+			if err != nil {
+				g.logger.Debugf("failed to get instance view: %v", err)
+				continue
+			}
+			if instanceView.BootDiagnostics != nil {
+				bootDiagnostics = append(bootDiagnostics, instanceView.BootDiagnostics)
+			}
+		}
+	}
+
+	return bootDiagnostics
+}
+
+func downloadFiles(ctx context.Context, fileURIs []string, credentials []*azblob.SharedKeyCredential, g *Gather) error {
+	var errs []error
+	var files []string
+
+	for _, fileURI := range fileURIs {
+		filePath, err := downloadFile(ctx, fileURI, credentials, g)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		files = append(files, filePath)
+	}
+
+	if len(files) > 0 {
+		err := gather.CreateArchive(files, g.serialLogBundle)
+		if err != nil {
+			g.logger.Debugf("failed to create archive: %s", err.Error())
+			errs = append(errs, err)
+		}
+	}
+
+	// clean up the mess we've made.
+	serialLogBundleDir := filepath.Join(g.directory, strings.TrimSuffix(filepath.Base(g.serialLogBundle), ".tar.gz"))
+	_, err := os.Stat(serialLogBundleDir)
+	if err == nil && !strings.HasPrefix(serialLogBundleDir, ".") {
+		err := os.RemoveAll(serialLogBundleDir)
+		if err != nil {
+			g.logger.Debugf("failed to remove archive directory: %s", err.Error())
+		}
+	}
```

Comment:
We are repeating this block of code in all of the providers. Maybe this could be part of the top-level gather like `CreateArchive`?

Reply:
Agreed.

Reply:
Fixed.